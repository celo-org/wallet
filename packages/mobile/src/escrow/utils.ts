import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToe164NumberSelector } from 'src/identity/reducer'
import { recipientCacheSelector } from 'src/recipients/reducer'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipientName = (payment: EscrowedPayment) => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToe164Number = useSelector(identifierToe164NumberSelector)
  const recipientCache = useSelector(recipientCacheSelector)
  const phoneNumber = identifierToe164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber]

  return recipient?.displayName || recipientPhone
}
