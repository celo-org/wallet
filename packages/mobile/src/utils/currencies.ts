export enum Currency {
  Celo = 'cGLD',
  Dollar = 'cUSD',
  Euro = 'cEUR',
}

export enum CiCoCurrency {
  CELO = 'CELO',
  CUSD = 'CUSD',
  CEUR = 'CEUR',
}

export interface CurrencyInfo {
  symbol: string
  displayDecimals: number
}

type CurrencyObject = { [key in Currency]: CurrencyInfo }

export type StableCurrency = Currency.Dollar | Currency.Euro
export const STABLE_CURRENCIES: StableCurrency[] = [Currency.Dollar, Currency.Euro]

export const CURRENCIES: CurrencyObject = {
  [Currency.Celo]: {
    symbol: '',
    displayDecimals: 3,
  },
  [Currency.Dollar]: {
    symbol: '$',
    displayDecimals: 2,
  },
  [Currency.Euro]: {
    symbol: 'â‚¬',
    displayDecimals: 2,
  },
}

export function mapOldCurrencyToNew(currencyString: string): Currency {
  const oldMapping: Record<string, any> = {
    dollar: Currency.Dollar,
    euro: Currency.Euro,
    gold: Currency.Celo,
  }
  const currency = oldMapping[currencyString]
  if (currency) {
    return currency
  }

  if (currencyString in Currency) {
    return currencyString as Currency
  }

  // Default value
  return Currency.Dollar
}
