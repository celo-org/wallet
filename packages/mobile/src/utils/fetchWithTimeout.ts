export const fetchWithTimeout = async (
  url: string,
  body: any | null,
  duration: number
): Promise<Response> => {
  try {
    const timeout = new Promise<undefined>((resolve, reject) => {
      const id = setTimeout(() => {
        clearTimeout(id)
        reject(Error(`Request timed out after ${duration}ms`))
      }, duration)
    })

    const response = await Promise.race([body ? fetch(url, body) : fetch(url), timeout])
    // Response should always be defined because `reject` throws an error
    // but just satifying the linter with this check
    if (!response) {
      throw Error(`Request timed out after ${duration}ms`)
    }
    return response
  } catch (error) {
    throw error
  }
}
