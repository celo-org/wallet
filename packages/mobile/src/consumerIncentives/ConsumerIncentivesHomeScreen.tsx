import Button, { BtnSizes } from '@celo/react-components/components/Button'
import Touchable from '@celo/react-components/components/Touchable'
import Times from '@celo/react-components/icons/Times'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import { StackScreenProps } from '@react-navigation/stack'
import BigNumber from 'bignumber.js'
import React, { Fragment } from 'react'
import { useAsync } from 'react-async-hook'
import { Trans, useTranslation } from 'react-i18next'
import { ActivityIndicator, Image, ScrollView, StyleSheet, Text, View } from 'react-native'
import { TouchableOpacity } from 'react-native-gesture-handler'
import { SafeAreaView, useSafeAreaInsets } from 'react-native-safe-area-context'
import { useDispatch } from 'react-redux'
import { showError } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { CELO_REWARDS_LINK } from 'src/brandingConfig'
import {
  ConsumerIncentivesData,
  fetchConsumerRewardsContent,
  Tier,
} from 'src/consumerIncentives/contentFetcher'
import i18n, { Namespaces } from 'src/i18n'
import { consumerIncentives, leaves } from 'src/images/Images'
import { noHeader } from 'src/navigator/Headers'
import { navigate, navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import useSelector from 'src/redux/useSelector'
import { stableTokenBalanceSelector } from 'src/stableToken/reducer'
import { getContentForCurrentLang } from 'src/utils/contentTranslations'
import Logger from 'src/utils/Logger'
import { formatFeedDate, getNextMondayAt8pmUtc } from 'src/utils/time'

const TAG = 'ConsumerIncentivesHomeScreen'

const useConsumerIncentivesContent = () => {
  const contentResult = useAsync<ConsumerIncentivesData>(fetchConsumerRewardsContent, [])
  let texts
  if (contentResult.result) {
    texts = getContentForCurrentLang(contentResult.result.content)
  }
  return {
    content: texts,
    tiers: contentResult.result?.tiers,
    loading: contentResult.loading,
    error: contentResult.error,
  }
}

const useAmountToAdd = (tiers: Tier[] | undefined) => {
  const dollarBalanceString = useSelector(stableTokenBalanceSelector) ?? '0'
  const dollarBalance = new BigNumber(dollarBalanceString)
  if (!tiers || tiers.length === 0) {
    return []
  }
  if (dollarBalance.isGreaterThanOrEqualTo(tiers[tiers.length - 1].minBalanceCusd)) {
    return []
  }
  let nextLevel = 1
  while (dollarBalance.isGreaterThanOrEqualTo(tiers[nextLevel - 1].minBalanceCusd)) {
    nextLevel++
  }
  return [
    new BigNumber(tiers[nextLevel - 1].minBalanceCusd).minus(dollarBalance).toFixed(0),
    nextLevel,
  ]
}

type Props = StackScreenProps<StackParamList, Screens.ConsumerIncentivesHomeScreen>
export default function ConsumerIncentivesHomeScreen(props: Props) {
  const { t } = useTranslation(Namespaces.consumerIncentives)
  const userIsVerified = useSelector((state) => state.app.numberVerified)
  const { content, tiers, loading, error } = useConsumerIncentivesContent()
  const [amountToAdd, nextLevel] = useAmountToAdd(tiers)
  const insets = useSafeAreaInsets()
  const dispatch = useDispatch()

  if (!loading && error) {
    Logger.error(TAG, 'Error while loading remote texts from Firebase', error)
    dispatch(showError(ErrorMessages.FIREBASE_FETCH_FAILED))
    navigateBack()
    return null
  }

  const onPressCTA = () => {
    if (userIsVerified) {
      navigate(Screens.FiatExchangeOptions, { isCashIn: true })
    } else {
      navigate(Screens.VerificationEducationScreen, { hideOnboardingStep: true })
    }
  }

  const onLearnMore = () => navigate(Screens.WebViewScreen, { uri: CELO_REWARDS_LINK })

  return (
    <ScrollView
      style={styles.container}
      contentContainerStyle={styles.scrollContainer}
      showsVerticalScrollIndicator={false}
    >
      <Image source={leaves} style={styles.backgroundImage} resizeMode={'stretch'} />
      <SafeAreaView edges={['bottom']} style={styles.contentContainer}>
        <Touchable
          style={[styles.closeButton, { marginTop: insets.top + 20 }]}
          onPress={navigateBack}
          borderless={true}
          hitSlop={variables.iconHitslop}
        >
          <Times />
        </Touchable>
        {loading && (
          <ActivityIndicator
            size="large"
            color={colors.greenBrand}
            style={styles.loading}
            testID="ConsumerIncentives/Loading"
          />
        )}
        {content && (
          <>
            <Image source={consumerIncentives} />
            <Text style={styles.title}>{t('title')}</Text>
            <Text style={[styles.body, styles.description]}>{t('description')}</Text>
            {amountToAdd && (
              <Text style={[styles.body, styles.description]}>
                <Trans
                  i18nKey={'addMore'}
                  ns={Namespaces.consumerIncentives}
                  tOptions={{
                    cUsdToAdd: amountToAdd,
                    date: formatFeedDate(getNextMondayAt8pmUtc(), i18n),
                    level: t('level', { level: nextLevel }),
                  }}
                >
                  <Text style={styles.bold} />
                  <Text style={styles.bold} />
                  <Text style={styles.bold} />
                </Trans>
              </Text>
            )}
            <View style={styles.tiersContainer}>
              {tiers &&
                tiers.map((tier, index) => (
                  <Fragment key={`tier${tier.celoReward}`}>
                    {index > 0 && <View style={styles.separator} />}
                    <Text style={styles.level}>{t('level', { level: index + 1 })}</Text>
                    <Text style={[styles.body, styles.tier]}>
                      {t('getCeloRewards', {
                        reward: tier.celoReward,
                        minBalance: tier.minBalanceCusd,
                      })}
                    </Text>
                  </Fragment>
                ))}
            </View>
            {content.extraSubtitle && <Text style={styles.subtitle}>{content.extraSubtitle}</Text>}
            {content.extraBody && <Text style={styles.body}>{content.extraBody}</Text>}
            {!userIsVerified && (
              <>
                <Text style={styles.subtitle}>{content.unverifiedSubtitle}</Text>
                <Text style={styles.body}>{content.unverifiedBody}</Text>
              </>
            )}
            <View style={styles.buttonContainer}>
              <Button
                size={BtnSizes.FULL}
                text={userIsVerified ? t('addCusd') : t('accountScreen10:confirmNumber')}
                onPress={onPressCTA}
                testID="ConsumerIncentives/CTA"
              />
            </View>
            <TouchableOpacity onPress={onLearnMore} testID="ConsumerIncentives/learnMore">
              <Text style={styles.learnMore}>{t('global:learnMore')}</Text>
            </TouchableOpacity>
          </>
        )}
      </SafeAreaView>
    </ScrollView>
  )
}

ConsumerIncentivesHomeScreen.navOptions = noHeader

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scrollContainer: {
    flexGrow: 1,
  },
  contentContainer: {
    flex: 1,
    alignItems: 'center',
    marginHorizontal: 24,
  },
  backgroundImage: {
    position: 'absolute',
    width: '100%',
    top: 0,
    left: 0,
    right: 0,
  },
  closeButton: {
    alignSelf: 'flex-start',
  },
  loading: {
    height: '100%',
  },
  title: {
    ...fontStyles.h2,
    marginTop: 16,
    textAlign: 'center',
  },
  description: {
    marginTop: 16,
  },
  tiersContainer: {
    width: '100%',
    marginTop: 32,
    padding: 16,
    borderWidth: 1,
    borderColor: colors.gray3,
    borderRadius: 4,
  },
  subtitle: {
    ...fontStyles.h2,
    fontSize: 20,
    marginTop: 40,
    marginBottom: 10,
    textAlign: 'center',
  },
  level: {
    ...fontStyles.regular,
    fontWeight: '600',
    textAlign: 'center',
    marginBottom: 8,
  },
  body: {
    ...fontStyles.regular,
    textAlign: 'center',
  },
  separator: {
    height: 1,
    alignSelf: 'stretch',
    backgroundColor: colors.gray3,
    marginVertical: 16,
    marginHorizontal: 8,
  },
  tier: {
    marginBottom: 4,
    paddingHorizontal: 12,
  },
  bold: {
    fontWeight: 'bold',
  },
  perMonth: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  buttonContainer: {
    marginTop: 36,
    width: '100%',
  },
  learnMore: {
    ...fontStyles.notificationHeadline,
    fontSize: 17,
    alignSelf: 'center',
    marginVertical: 24,
    color: colors.greenUI,
  },
})
