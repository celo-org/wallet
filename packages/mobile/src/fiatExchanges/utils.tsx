import {
  CurrencyCode,
  DEFAULT_TESTNET,
  IP_ADDRESS_LOCATION_URL_PROD,
  IP_ADDRESS_LOCATION_URL_STAGING,
  PROVIDER_URL_COMPOSER_PROD,
  PROVIDER_URL_COMPOSER_STAGING,
  SIMPLEX_URI,
} from 'src/config'
import { Providers } from 'src/fiatExchanges/ProviderOptionsScreen'
import { providerAvailability } from 'src/flags'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { navigateToURI } from 'src/utils/linking'

interface RequestData {
  address: string | null
  digitalAsset: string
  fiatCurrency: string
  fiatAmount: number
}
export interface UserLocation {
  country: string | null
  state: string | null
}

export const fetchProviderWidgetUrl = async (provider: Providers, requestData: RequestData) => {
  const response = await fetch(
    DEFAULT_TESTNET === 'mainnet' ? PROVIDER_URL_COMPOSER_PROD : PROVIDER_URL_COMPOSER_STAGING,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...requestData,
        provider,
        env: DEFAULT_TESTNET === 'mainnet' ? 'production' : 'staging',
      }),
    }
  )

  return response.json()
}

export const fetchUserIpAddress = async () => {
  const response = await fetch(
    DEFAULT_TESTNET === 'mainnet' ? IP_ADDRESS_LOCATION_URL_PROD : IP_ADDRESS_LOCATION_URL_STAGING,
    {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        urlType: 'ip',
        env: DEFAULT_TESTNET === 'mainnet' ? 'production' : 'staging',
      }),
    }
  )

  return response.json()
}

export const isExpectedUrl = (fetchedUrl: string, providerUrl: string) =>
  fetchedUrl.startsWith(providerUrl)

export const openMoonpay = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.MoonPayScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

export const openSimplex = (account: string | null) => {
  navigateToURI(`${SIMPLEX_URI}?address=${account}`)
}

export const openRamp = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.RampScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

export const openTransak = (
  amount: number,
  currencyCode: LocalCurrencyCode,
  currencyToBuy: CurrencyCode
) => {
  navigate(Screens.TransakScreen, {
    localAmount: amount,
    currencyCode,
    currencyToBuy,
  })
}

type ProviderAvailability = typeof providerAvailability
type SpecificProviderAvailability = { [K in keyof ProviderAvailability]: boolean }

type Entries<T> = Array<{ [K in keyof T]: [K, T[K]] }[keyof T]>

export function getProviderAvailability(
  userLocation: UserLocation | undefined
): SpecificProviderAvailability {
  // tslint:disable-next-line: no-object-literal-type-assertion
  const { countryCodeAlpha2, stateCode } = userLocation
    ? { countryCodeAlpha2: userLocation.country, stateCode: userLocation.state }
    : { countryCodeAlpha2: null, stateCode: null }

  const features = {} as SpecificProviderAvailability
  for (const [key, value] of Object.entries(providerAvailability) as Entries<
    ProviderAvailability
  >) {
    if (!countryCodeAlpha2) {
      features[key] = false
    } else {
      if (countryCodeAlpha2 === 'US' && (value as any)[countryCodeAlpha2] !== true) {
        features[key] = stateCode ? (value as any)[countryCodeAlpha2][stateCode] ?? false : false
      } else {
        features[key] = (value as any)[countryCodeAlpha2] ?? false
      }
    }
  }
  return features
}
