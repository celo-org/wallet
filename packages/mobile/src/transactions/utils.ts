import { TokenTransactionType } from 'src/apollo/types'
import { ExchangeConfirmationCardProps } from 'src/exchange/ExchangeConfirmationCard'
import { CURRENCIES, CURRENCY_ENUM } from 'src/geth/consts'
import i18n from 'src/i18n'
import { AddressToDisplayNameType } from 'src/identity/reducer'
import { FeedItem } from 'src/transactions/TransactionFeed'
import { TransferConfirmationCardProps } from 'src/transactions/TransferConfirmationCard'
import { formatFeedSectionTitle, timeDeltaInDays } from 'src/utils/time'

// Groupings:
// Recent -> Last 7 days.
// [Current month] - "July" -> Captures transactions from the current month that aren’t captured in Recent.
// [Previous months] - "June" -> Captures transactions by month.
// [Months over a year ago] — "July 2019" -> Same as above, but with year appended.
// Sections are hidden if they have no items.
export const groupFeedItemsInSections = (feedItems: FeedItem[]) => {
  const sectionsMap: {
    [key: string]: {
      data: FeedItem[]
      daysSinceTransaction: number
    }
  } = {}

  feedItems.reduce((sections, item) => {
    const daysSinceTransaction = timeDeltaInDays(Date.now(), item.timestamp)
    const key =
      daysSinceTransaction <= 7
        ? i18n.t('walletFlow5:feedSectionHeaderRecent')
        : formatFeedSectionTitle(item.timestamp, i18n)
    sections[key] = sections[key] || {
      daysSinceTransaction,
      data: [],
    }
    sections[key].data.push(item)
    return sections
  }, sectionsMap)

  return Object.entries(sectionsMap)
    .sort((a, b) => a[1].daysSinceTransaction - b[1].daysSinceTransaction)
    .map(([key, value]) => ({
      title: key,
      data: value.data,
    }))
}

export const exchangeReviewHeader = (confirmationProps: ExchangeConfirmationCardProps) => {
  const { makerAmount } = confirmationProps
  const isSold = makerAmount.currencyCode === CURRENCIES[CURRENCY_ENUM.GOLD].code
  return isSold ? i18n.t('exchangeFlow9:soldGold') : i18n.t('exchangeFlow9:purchasedGold')
}

export const transferReviewHeader = (
  type: TokenTransactionType,
  confirmationProps: TransferConfirmationCardProps,
  addressToDisplayName: AddressToDisplayNameType
) => {
  let headerText = ''
  const isCeloTx = confirmationProps.amount.currencyCode === CURRENCIES[CURRENCY_ENUM.GOLD].code
  switch (type) {
    case TokenTransactionType.Sent:
      headerText = i18n.t(
        isCeloTx ? 'walletFlow5:transactionHeaderWithdrewCelo' : 'walletFlow5:transactionHeaderSent'
      )
      break
    case TokenTransactionType.EscrowSent:
      headerText = i18n.t('walletFlow5:transactionHeaderEscrowSent')
      break
    case TokenTransactionType.Received:
      if (addressToDisplayName[confirmationProps.address || '']?.isCeloRewardSender) {
        headerText = i18n.t('walletFlow5:transactionHeaderCeloReward')
      }
      headerText = isCeloTx
        ? i18n.t('walletFlow5:transactionHeaderCeloDeposit')
        : i18n.t('walletFlow5:transactionHeaderReceived')
      break
    case TokenTransactionType.EscrowReceived:
      headerText = i18n.t('walletFlow5:transactionHeaderEscrowReceived')
      break
    case TokenTransactionType.VerificationFee:
      headerText = i18n.t('walletFlow5:transactionHeaderVerificationFee')
      break
    case TokenTransactionType.Faucet:
      headerText = i18n.t('walletFlow5:transactionHeaderFaucet')
      break
    case TokenTransactionType.InviteSent:
      headerText = i18n.t('walletFlow5:transactionHeaderInviteSent')
      break
    case TokenTransactionType.InviteReceived:
      headerText = i18n.t('walletFlow5:transactionHeaderInviteReceived')
      break
    case TokenTransactionType.NetworkFee:
      headerText = i18n.t('walletFlow5:transactionHeaderNetworkFee')
      break
  }

  return headerText
}
