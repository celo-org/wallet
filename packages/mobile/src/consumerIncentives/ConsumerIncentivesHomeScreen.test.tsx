import * as React from 'react'
import 'react-native'
import { fireEvent, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import { CELO_REWARDS_LINK } from 'src/config'
import ConsumerIncentivesHomeScreen from 'src/consumerIncentives/ConsumerIncentivesHomeScreen'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

const mockScreenProps = getMockStackScreenProps(Screens.ConsumerIncentivesHomeScreen)

const createStore = (numberVerified: boolean) =>
  createMockStore({
    app: { numberVerified },
    stableToken: { balance: '1000' },
  })

describe('ConsumerIncentivesHomeScreen', () => {
  beforeEach(() => jest.useRealTimers())

  it('renders correctly', async () => {
    const tree = render(
      <Provider store={createStore(true)}>
        <ConsumerIncentivesHomeScreen {...mockScreenProps} />
      </Provider>
    )
    expect(tree.queryByTestId('ConsumerIncentives/CTA')).toBeTruthy()
    expect(tree.queryByTestId('ConsumerIncentives/learnMore')).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('navigates to cash in screen if user is verified and CTA is tapped', async () => {
    const { getByTestId } = render(
      <Provider store={createStore(true)}>
        <ConsumerIncentivesHomeScreen {...mockScreenProps} />
      </Provider>
    )

    fireEvent.press(getByTestId('ConsumerIncentives/CTA'))
    expect(navigate).toHaveBeenCalledWith(Screens.FiatExchangeOptions, { isCashIn: true })
  })

  it('navigates to Phone Confirmation screen if user is not verified and CTA is tapped', async () => {
    const { getByTestId } = render(
      <Provider store={createStore(false)}>
        <ConsumerIncentivesHomeScreen {...mockScreenProps} />
      </Provider>
    )
    fireEvent.press(getByTestId('ConsumerIncentives/CTA'))

    expect(navigate).toHaveBeenCalledWith(Screens.VerificationEducationScreen, {
      hideOnboardingStep: true,
    })
  })

  it('opens a WebView when Learn More is tapped', async () => {
    const { getByTestId } = render(
      <Provider store={createStore(true)}>
        <ConsumerIncentivesHomeScreen {...mockScreenProps} />
      </Provider>
    )
    fireEvent.press(getByTestId('ConsumerIncentives/learnMore'))

    expect(navigate).toHaveBeenCalledWith(Screens.WebViewScreen, {
      uri: CELO_REWARDS_LINK,
    })
  })
})
