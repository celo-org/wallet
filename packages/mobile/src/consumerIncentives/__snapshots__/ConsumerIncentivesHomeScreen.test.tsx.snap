// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsumerIncentivesHomeScreen renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  showsVerticalScrollIndicator={false}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <Image
      resizeMode="stretch"
      source={
        Object {
          "testUri": "../../../packages/mobile/src/images/leaves.png",
        }
      }
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": "100%",
        }
      }
    />
    <RNCSafeAreaView
      edges={
        Array [
          "bottom",
        ]
      }
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignSelf": "flex-start",
            },
            Object {
              "marginTop": 20,
            },
          ]
        }
      >
        <svg
          height={16}
          style={Object {}}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13.9999 2.00146L1.99994 14.0015"
            stroke="#2E3338"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            style={Object {}}
          />
          <path
            d="M1.99994 2.00146L13.9999 14.0015"
            stroke="#2E3338"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth={2}
            style={Object {}}
          />
        </svg>
      </View>
      <Image
        source={
          Object {
            "testUri": "../../../packages/mobile/src/images/consumerIncentives.png",
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Jost-Medium",
            "fontSize": 22,
            "lineHeight": 28,
            "marginTop": 16,
            "textAlign": "center",
          }
        }
      >
        title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
              "textAlign": "center",
            },
            Object {
              "marginTop": 16,
            },
          ]
        }
      >
        description
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "lineHeight": 22,
              "textAlign": "center",
            },
            Object {
              "marginTop": 16,
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        />
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        />
        <Text
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        />
      </Text>
      <View
        style={
          Object {
            "borderColor": "#B4B9BD",
            "borderRadius": 4,
            "borderWidth": 1,
            "marginTop": 32,
            "padding": 16,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 22,
              "marginBottom": 8,
              "textAlign": "center",
            }
          }
        >
          level
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "textAlign": "center",
              },
              Object {
                "marginBottom": 4,
                "paddingHorizontal": 12,
              },
            ]
          }
        >
          getCeloRewards
        </Text>
        <View
          style={
            Object {
              "alignSelf": "stretch",
              "backgroundColor": "#B4B9BD",
              "height": 1,
              "marginHorizontal": 8,
              "marginVertical": 16,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 22,
              "marginBottom": 8,
              "textAlign": "center",
            }
          }
        >
          level
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "textAlign": "center",
              },
              Object {
                "marginBottom": 4,
                "paddingHorizontal": 12,
              },
            ]
          }
        >
          getCeloRewards
        </Text>
        <View
          style={
            Object {
              "alignSelf": "stretch",
              "backgroundColor": "#B4B9BD",
              "height": 1,
              "marginHorizontal": 8,
              "marginVertical": 16,
            }
          }
        />
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 22,
              "marginBottom": 8,
              "textAlign": "center",
            }
          }
        >
          level
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
                "textAlign": "center",
              },
              Object {
                "marginBottom": 4,
                "paddingHorizontal": 12,
              },
            ]
          }
        >
          getCeloRewards
        </Text>
      </View>
      <View
        style={
          Object {
            "marginTop": 36,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexDirection": "column",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 100,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#1AB775",
                  "flexGrow": 1,
                  "height": 48,
                  "justifyContent": "center",
                  "opacity": undefined,
                  "paddingHorizontal": 24,
                  "paddingVertical": 5,
                }
              }
              testID="ConsumerIncentives/CTA"
            >
              <Text
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                addCusd
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RNGestureHandlerButton
        collapsable={false}
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        testID="ConsumerIncentives/learnMore"
      >
        <View
          accessible={true}
          style={
            Array [
              Object {},
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "#1AB775",
                "fontFamily": "Jost-Medium",
                "fontSize": 17,
                "lineHeight": 22,
                "marginVertical": 24,
              }
            }
          >
            global:learnMore
          </Text>
        </View>
      </RNGestureHandlerButton>
    </RNCSafeAreaView>
  </View>
</RCTScrollView>
`;
